service/frontend/adminHandler.go

```diff
@@ -421,6 +423,237 @@ func (adh *AdminHandler) GetWorkflowExecutionRawHistory(
+	versionHistories *persistence.VersionHistories,
+) (int64, int64, *persistence.VersionHistory, error) {
+
+	currentBranch, err := versionHistories.GetCurrentVersionHistory()
+	// The start event should be exclusive
+	targetStartEventID := request.GetStartEventId() + 1
+	targetStartEventVersion := request.GetStartEventVersion()
+	if request.StartEventId == nil || request.StartEventVersion == nil {
+		firstItem, err := currentBranch.GetFirstItem()
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		targetStartEventID = common.FirstEventID
+		targetStartEventVersion = firstItem.GetVersion()
+	}
+	targetEndEventID := request.GetEndEventId()
+	targetEndEventVersion := request.GetEndEventVersion()
+	if request.EndEventId == nil || request.EndEventVersion == nil {
+		lastItem, err := currentBranch.GetLastItem()
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		targetEndEventID = lastItem.GetEventID()
+		targetEndEventVersion = lastItem.GetVersion()
+	}
+
+	// update current branch based on updated end event item
+	targetItem := persistence.NewVersionHistoryItem(targetEndEventID, targetEndEventVersion)
+	idx, err := versionHistories.FindFirstVersionHistoryIndexByItem(targetItem)
+	if err != nil {
+		return 0, 0, nil, err
+	}
+	currentBranch, err = versionHistories.GetVersionHistory(idx)
+	if err != nil {
+		return 0, 0, nil, err
+	}
+
+	startItem := persistence.NewVersionHistoryItem(targetStartEventID, targetStartEventVersion)
+	// Start event is not on the same branch as target branch
+	if !currentBranch.ContainsItem(startItem) {
+		idx, err := versionHistories.FindFirstVersionHistoryIndexByItem(startItem)
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		startBranch, err := versionHistories.GetVersionHistory(idx)
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		startItem, err = currentBranch.FindLCAItem(startBranch)
+		if err != nil {
+			return 0, 0, nil, err
+		}
+		targetStartEventID = startItem.GetEventID()
+	}
+
+	return targetStartEventID, targetEndEventID, currentBranch, nil
+}
+
+func (adh *AdminHandler) preparePaginationToken(
+	ctx context.Context,
+	request *admin.GetWorkflowExecutionRawHistoryV2Request,
+	branchToken []byte,
+	startEventID int64,
+	endEventID int64,
+) (*getHistoryContinuationToken, error) {
+
+	execution := request.Execution
+	var token *getHistoryContinuationToken
+	if request.NextPageToken != nil {
+		token, err := deserializeHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+
+		if execution.GetRunId() != token.RunID ||
+			// we guarantee to use the first event ID provided in the request
+			request.GetStartEventId() != token.FirstEventID ||
+			// the next event ID in the request must be <= next event ID from mutable state, when initialized
+			// so as long as customer do not change next event ID during pagination,
+			// next event ID in the token <= next event ID in the request.
+			request.GetEndEventId() < token.NextEventID {
```

Comment:
this API aims to dump history in certain range, so run ID, start event ID / version, end event ID / version should not change at all.

the check here is not enough.

I would also suggest making a new pagination token for this use case

Reply:
Make a new token for this API